Add CUDA and ROCM docs on website (#8172)
google/iree: docs/website/docs/deployment-configurations/gpu-cuda-rocm.md

# CUDA and ROCM GPU HAL Driver

> A lot of AI/ML workload uses GPU to accelerate execution. IREE can accelerate model execution on GPUs through CUDA for NVIDIA hardwares, and ROCm for AMD hardwares. Due to the similarity of CUDA and ROCM APIs and infrastructure, CUDA and ROCM backend shares many features and implementation in IREE. In IREE compiler level, CUDA and ROCM bothg go through LLVMGPU pipeline, but CUDA generate ptx, while ROCM generate hsaco. In the IREE runtime level, the HAL layer for ROCM mirrors the one on CUDA, except for the command graph. Where CUDA has direct, stream, and graph command buffer, and ROCM has only direct command buffer.

Reviewer: a few grammar and capitalization nits

```suggestion
IREE can accelerate model execution on NVIDIA GPUs using CUDA and on AMD GPUs using ROCm. Due to the similarity of CUDA and ROCm APIs and infrastructure, the CUDA and ROCm backends share much of their implementation in IREE. The IREE compiler uses a similar GPU code generation pipeline for each, but generates PTX for CUDA and hsaco for ROCm. The IREE runtime HAL driver for ROCm mirrors the one for CUDA, except for the command graph - where CUDA has "direct", "stream", and "graph" command buffers, and ROCM has only "direct" command buffers.
```

## Prerequisites

> In order to use CUDA or ROCM to drive the GPU, you need to have a functional CUDA or ROCM

Reviewer: nit: ROCm

```suggestion
In order to use CUDA or ROCm to drive the GPU, you need to have a functional CUDA or ROCm
```

platforms.

=== "Nvidia/CUDA"

Ensure that the `IREE_TARGET_BACKEND_CUDA` CMake option is `ON` when
configuring for the host.

=== "AMD/ROCM"

Ensure that the `IREE_TARGET_BACKEND_ROCM` CMake option is `ON` when
configuring for the host.

## Compile and run the model

With the compiler and runtime for CUDA ready, we can now compile a model
and run it on the GPU.

### Compile the model

IREE compilers transform a model into its final deployable format in many
sequential steps. A model authored with Python in an ML framework should use the
corresponding framework's import tool to convert into a format (i.e.,
[MLIR][mlir]) expected by main IREE compilers first.

Using MobileNet v2 as an example, you can download the SavedModel with trained
weights from [TensorFlow Hub][tf-hub-mobilenetv2] and convert it using IREE's
[TensorFlow importer][tf-import]. Then,

#### Compile using the command-line

Let `iree_input.mlir` be the model's initial MLIR representation generated by
IREE's TensorFlow importer. We can now compile them for each GPU by running the following command:

=== "Nvidia/CUDA"

> ``` shell hl_lines="3 4"

Reviewer: ```suggestion
``` shell hl_lines="3 5"
```

highlight line 5 also?

this is how the highlighting appears on the site:
![image](https://user-images.githubusercontent.com/4010439/150881147-55c8bb7b-314e-490e-8570-368c36e09f74.png)

docs: https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#highlighting-specific-lines

iree/tools/iree-translate \
    -iree-mlir-to-vm-bytecode-module \
    -iree-hal-target-backends=cuda \
    -iree-cuda-llvm-target-arch=<...> \
    -iree-hal-cuda-disable-loop-nounroll-wa \
    iree_input.mlir -o mobilenet-cuda.vmfb
```

Note that a cuda target architecture(`iree-cuda-llvm-target-arch`) of the form `sm_<arch_number>` is needed
to compile towards each GPU architecture. If no architecture is specified then we will default to `sm_35`
Here are a table of commonly used architecture

CUDA GPU  | Target Architecture
:--------: | :-----------:
Nvidia K80 | `sm_35`
Nvidia P100 | `sm_60`
Nvidia V100 | `sm_70`
Nvidia A100 | `sm_80`

=== "AMD/ROCM"

> ``` shell hl_lines="3 4"
> iree/tools/iree-translate \
>     -iree-mlir-to-vm-bytecode-module \
>     -iree-hal-target-backends=rocm \
>     -iree-rocm-target-chip=<...> \
>     -iree-rocm-link-bc=true \
>     -iree-rocm-bc-dir=<...> \

Reviewer: highlight all rocm lines?

```suggestion
``` shell hl_lines="3 6"
iree/tools/iree-translate \
    -iree-mlir-to-vm-bytecode-module \
    -iree-hal-target-backends=rocm \
    -iree-rocm-target-chip=<...> \
    -iree-rocm-link-bc=true \
    -iree-rocm-bc-dir=<...> \
```

    iree_input.mlir -o mobilenet-rocm.vmfb
```

> Note ROCM Bitcode Dir(`iree-rocm-bc-dir`) path is required. If the system you are compiling IREE in has ROCM installed, then the default value of `/opt/rocm/amdgcn/bitcode` will usually suffice. If you intend on building ROCM compiler in a non-ROCM capable system, please set `iree-rocm-bc-dir` to the absolute path where you might have saved the amdgcn bitcode.

Reviewer: nit: ROCm
```suggestion
Note ROCm Bitcode Dir(`iree-rocm-bc-dir`) path is required. If the system you are compiling IREE in has ROCm installed, then the default value of `/opt/rocm/amdgcn/bitcode` will usually suffice. If you intend on building ROCm compiler in a non-ROCm capable system, please set `iree-rocm-bc-dir` to the absolute path where you might have saved the amdgcn bitcode.
```

Note that a rocm target chip(`iree-rocm-target-chip`) of the form `gfx<arch_number>` is needed
to compile towards each GPU architecture. If no architecture is specified then we will default to `gfx908`
Here are a table of commonly used architecture

AMD GPU  | Target Chip
:--------: | :-----------:
AMD MI25 | `gfx900`
AMD MI50 | `gfx906`
AMD MI60 | `gfx906`
AMD MI100 | `gfx908`

### Run the model

#### Run using the command-line

In the build directory, run the following command:

=== "Nvidia/CUDA"

``` shell hl_lines="2"
iree/tools/iree-run-module \
    --driver=cuda \
    --module_file=mobilenet-cuda.vmfb \
    --entry_function=predict \
    --function_input="1x224x224x3xf32=0"
```

=== "AMD/ROCM"

``` shell hl_lines="2"
iree/tools/iree-run-module \

>     --driver=cuda \

Reviewer: should this use `--driver=rocm`?
```suggestion
    --driver=rocm \
```
